scalabel/eval/result.py

```diff
@@ -0,0 +1,167 @@
+"""Definition of the base evaluation result class."""
+
+from collections import defaultdict
+from typing import AbstractSet, Callable, Dict, List, Optional, Union
+
+import numpy as np
+from mypy_extensions import KwArg, VarArg
+from pandas import DataFrame
+from pydantic import BaseModel, PrivateAttr
+
+FORMATTER = Callable[[VarArg(object), KwArg(object)], str]
+Scores = Dict[str, Union[int, float]]
+ScoresList = List[Scores]
+AVERAGE = "AVERAGE"
+OVERALL = "OVERALL"
+
+
+class Result(BaseModel):
+    """The base class for bdd100k evaluation results.
+
+    Each data field corresponds to a evluation metric. The value for each
+    metric is a dict that maps the category names to scores.
+
+    Functions:
+        dict() -> dict[str, dict[str, int | float]]:
+            export all data to a nested dict.
+        json() -> str:
+            export the nested dict of `dict()` into a JSON string.
+        pd_frame() -> pandas.DataFrame:
+            export data fields to a formatted DataFrame.
+        table() -> str:
+            export data fields to a formatted table string.
+        summary() -> dict[str, int | float]:
+            export most important fields to a flattened dict.
+        __str__() -> str:
+            the same as `table()`.
+    """
+
+    _formatters: Dict[str, FORMATTER] = PrivateAttr(dict())
+    _row_breaks: List[int] = PrivateAttr([])
+
+    def __init__(self, **data: Union[int, float, ScoresList]) -> None:
+        """Check the input structure and initiliaze the model."""
+        data_check: Dict[str, ScoresList] = {
+            metric: cont
+            for metric, cont in data.items()
+            if isinstance(cont, list)
+        }
+        ref_scores_list = data_check[list(data_check.keys())[0]]
```

Comment:
if we check this here, we should add a remark to the doc that all keys need to be the same for the different metrics